primitives: Improve `KeyNibbles` serialization

jsdanielh opened the pull request:
- Improve `KeyNibbles` serialization performance by using`serde_byte`s `ByteBuf`.
- Commit e983d05 introduced a `TrieNode` serialization performance improvement by using `serde_byte`'s `Bytes` but forgot to also use it for deserizalization. But the problem is that we're deserializing into a `Vec<u8>` and we can't use `Bytes` since the size is unknown. So this commit changes (de)serialization to use `ByteBuf` instead.

## Pull request checklist

- [X] All tests pass. The project builds and runs.
- [X] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

jsdanielh commented:
@hrxi Per the profiling I did, this improves a lot the serialization of `KeyNibbles` but now we're potentially storing more data than needed. Any idea on how to improve it?

hrxi reviewed (changes requested):
You could probably gain the same performance by using [`ByteBuf`](https://docs.rs/serde_bytes/0.11.15/serde_bytes/struct.ByteBuf.html) instead of `Vec` for de-/serialization, without changing the on-disk format.

jsdanielh committed:
primitives: Improve KeyNibbles serialization

Improve `KeyNibbles` serialization performance by using`serde_byte`s
`ByteBuf`.

jsdanielh committed:
primitives: Consistently deserialize a `TrieNode`

Commit e983d05 introduced a `TrieNode` serialization performance
improvement by using `serde_byte`'s `Bytes` but forgot to also use it
for deserizalization. But the problem is that we're deserializing
into a `Vec<u8>` and we can't use `Bytes` since the size is unknown.
So this commit changes (de)serialization to use `ByteBuf` instead.

hrxi commented:
Added a9a1d370d098c1e8f91bd0c5f54da4ce6687fa9b to #2766.
